sled-agent/src/rack_setup/service.rs

```diff
@@ -988,6 +1006,119 @@ impl ServiceInner {
+                .map_err(|e| {
+                    SetupServiceError::BadConfig(format!(
+                    "unable to parse rack_network_config.uplink_up as IpAddr: {e}"))
+                })?;
+
+            let link_settings = LinkSettings {
+                // TODO Allow user to configure link properties
+                // https://github.com/oxidecomputer/omicron/issues/3061
+                params: LinkCreate {
+                    autoneg: false,
+                    kr: false,
+                    fec: PortFec::None,
+                    speed: rack_network_config.uplink_port_speed.clone().into(),
+                },
+                addrs: vec![addr],
+            };
+
+            dpd_port_settings.links.insert(link_id.to_string(), link_settings);
+
+            let port_id: PortId = rack_network_config.uplink_port.parse()
+                .map_err(|e| SetupServiceError::BadConfig(
+                        format!("could not use value provided to rack_network_config.uplink_port as PortID: {e}")
+                ))?;
+
+            let nexthop: Option<Ipv4Addr> = Some(rack_network_config.gateway_ip.parse()
+                .map_err(|e| SetupServiceError::BadConfig(
+                        format!("unable to parse rack_network_config.gateway_ip as Ipv4Addr: {e}")
+                ))?);
+
+            dpd_port_settings.v4_routes.insert(
+                Ipv4Cidr { prefix: "0.0.0.0".parse().unwrap(), prefix_len: 0 }
+                    .to_string(),
+                RouteSettingsV4 { link_id: link_id.0, nexthop },
+            );
+
+            info!(self.log, "Waiting for dendrite to come online");
+            while dpd.dpd_uptime().await.is_err() {
+                tokio::time::sleep(std::time::Duration::from_secs(2)).await;
+            }
+
+            info!(self.log, "Configuring boundary services loopback address on switch"; "config" => format!("{body:#?}"));
+            dpd.loopback_ipv6_create(&body).await.map_err(|e| {
+                SetupServiceError::Dendrite(format!(
+                    "unable to create inital switch loopback address: {e}"
+                ))
+            })?;
+
+            info!(self.log, "Configuring default uplink on switch"; "config" => format!("{dpd_port_settings:#?}"));
+            dpd.port_settings_apply(&port_id, &dpd_port_settings)
+                .await
+                .map_err(|e| {
+                    SetupServiceError::Dendrite(format!(
+                    "unable to apply initial uplink port configuration: {e}"
+                ))
+                })?;
+
+            info!(self.log, "advertising boundary services loopback address");
+            let mut ddmd_addr = dpd_addr;
+            ddmd_addr.set_port(8000);
+            let ddmd_client = DdmAdminClient::new(&self.log, ddmd_addr)?;
+            ddmd_client.advertise_prefix(Ipv6Subnet::new(
+                BOUNDARY_SERVICES_ADDR.parse().unwrap(),
+            ));
```

Comment:
Are these calls all idempotent?

Reply:
This ddm call for advertising prefixes is idempotent.